keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,842 @@
+)
+
+type PodSpec struct {
+...
+  // UserNamespaceMode controls how user namespaces are used for this Pod.
+  // Three modes are supported:
+  // "Host": The pod shares the host user namespace. (default value).
+  // "Cluster": The pod uses a cluster-wide configured ID mappings.
+  // +k8s:conversion-gen=false
+  // +optional
+  UserNamespaceMode PodUserNamespaceMode `json:"userNamespaceMode,omitempty" protobuf:"bytes,36,opt  name=userNamespaceMode"`
+...
+```
+
+#### Option 2: PodSecurityContext
+
+Add it to `v1.PodSpec.PodSecurityContext` as this field controls a security
+aspect of the pod.
+
+```
+const (
+	UserNamespaceModeHost    PodUserNamespaceMode = "Host"
+	UserNamespaceModeCluster PodUserNamespaceMode = "Cluster"
+)
+
+type PodSecurityContext struct {
+...
+  // UserNamespaceMode controls how user namespaces are used for this Pod.
+  // Three modes are supported:
+  // "Host": The pod shares the host user namespace. (default value).
+  // "Cluster": The pod uses a cluster-wide configured ID mappings.
+  // +optional
+  UserNamespaceMode PodUserNamespaceMode `json:"userNamespaceMode,omitempty" protobuf:"bytes,11,opt  name=userNamespaceMode"`
+...
+```
+<<[/UNRESOLVED]>>
+
+### Configuring the Cluster ID Mappings
+
+This proposal considers two different ways to configure the ID mappings used for
+the `Cluster` mode. This is for discussion with the community and only one will
+be considered.
+
+<<[UNRESOLVED where to configure the cluster wide ID mappings ]>>
+
+#### Option 1: Configure in Kubelet Configuration File
+
+The ID mappings used for pods in `Cluster` mode are set in the kubelet
+configuration file. The file allows to set different ID mappings for user and
+group. Kubelet performs a check looking for configuration mistakes, like
+overlapping mappings, to prevent kubelet sending wrong ID mappings to the
+runtime.
+
+The following is an example a configuration file:
+
+```
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: KubeletConfiguration
+clusterIDMappings:
+  uidMappings:
+  - containerID: 0
+    hostID: 500000
+    size: 65536
+  - containerID: 100000
+    hostID: 600000
+    size: 65536
+  gidMappings:
+  - containerID: 0
+    hostID: 500000
+    size: 65536
+
+```
+
+**Pros**:
+ - It's directly accessible by the kubelet. It's not needed to expose through an
+   API and there are not concerns if the api server is down.
+ - No modifications are needed to other components.
+
+**Cons**:
+ - It's more difficult for the user as they have to guarantee the same ID
+   mapping is configured in all the nodes.
+
+#### Option 2: Configure as a Cluster Parameter in kube-apiserver
+
+This option considers setting this parameter on the kube-apiserver.
+
+**Pros**:
+ - Easier for the user as the parameter is configured once in a single place.
+
+ **Cons**:
+ - It's difficult to expose this parameter to the kubelet.
+ - The parameter could not be available for the kubelet if the kube-apiserver is
+   down.
+
```

Comment:
Option 3 - stick it in Node
Option 4 - make a new resource which is 1:1 with Nodes for this sort of info